eclass/linux-mod-r1.eclass

```diff
@@ -545,10 +557,21 @@ modules_post_process() {
 	(( ${#mods[@]} )) ||
 		die "${FUNCNAME[0]} was called with no installed modules under ${path}"
 
-	# TODO?: find way for sane use with dracut (its 90kernel-modules-extra
-	# parses depmod.d files directly and assumes should include its modules
-	# which can lead to unnecessarily increased size or stale modules)
-#	_modules_process_depmod.d "${mods[@]#"${path}/"}"
+	# Instruct Dracut on whether or not we want these modules in initramfs
+	insinto /etc/dracut.conf.d
+	local dracut_arg base_name module_names=()
+	if use initramfs; then
+		dracut_arg=add_drivers
+	else
+		dracut_arg=omit_drivers
+	fi
+	for mod in "${mods[@]}"; do
+		base_name=${mod##*/}
+		module_names+=( ${base_name%.ko} )
+	done
+	newins - 10-${PN}.conf <<-EOF || die
+		${dracut_arg}="${module_names[@]}"
```

Comment:
Another thing just realized, these add_drivers files may be insufficient, e.g. if someone included nvidia-drivers they'd be missing the gsp firmware files.

Reply:
To make things more annoying, not everyone need these and they are an extra 60MB. Given including nvidia-drivers is discouraged and people really wanted that were doing it on their own I didn't have to worry about it.

Could always just mask the USE if wanted it to stay that way though.

Reply:
I thought dracut was smart enough to figure out which firmware files it needs for which drivers.

If that is not the case then we'd either have to make some mapping of drivers to firmware files in the ebuilds where required, or leave firmware files up to local user configuration.

Reply:
> I thought dracut was smart enough to figure out which firmware files it needs for which drivers.

I have tested this now, and dracut is indeed smart enough to figure out what firmware files we want based on the included drivers and the host_only setting. So we should have no problem with the firmware files.

Reply:
Was under the impression it looked at loaded firmwares or something, how is it finding that nvidia-drivers uses a firmware? I can't find generic code that would handle this.

That doesn't seem to be stored anywhere, is it grepping the module?

Reply:
(not a big deal even if it doesn't work though, still don't intend to overly care for the nvidia-on-initramfs situation)

Reply:
...albeit if it can result in display not working due to missing firmware, that's pretty much a confirmed mask though unless bother adding custom handling for it (nvidia+firmware = ~100MB that don't compress that well for a big initramfs too)

Reply:
> Was under the impression it looked at loaded firmwares or something, how is it finding that nvidia-drivers uses a firmware? I can't find generic code that would handle this.
> 
> That doesn't seem to be stored anywhere, is it grepping the module?

I think this is responsible for finding which firmware to install for a given module: https://github.com/dracut-ng/dracut-ng/blob/fd384f31599ff231f5a41d876d473d6ac9bb4144/src/install/dracut-install.c#L1399
It's parsing modinfo.

> ...albeit if it can result in display not working due to missing firmware, that's pretty much a confirmed mask though unless bother adding custom handling for it (nvidia+firmware = ~120MB that don't compress that well for a big initramfs too)

It should in principle be possible for ebuilds to install additional dracut.conf.d snippets based on the value of the initramfs flag. But given the above, I don't think it is required in this case since `sys-kernel/linux-firmware` also rebuilds the initramfs, so even if we install first nvidia-drivers and then linux-firmware we still end up with an initramfs that contains both the kernel modules and the accompanying firmware. We could even add `RDEPEND="initramfs? ( sys-kernel/linux-firmware )"` to enforce that we always have linux-firmware before nvidia-drivers.

Or, given that the whole nvidia framebuffer is an experimental feature, we can indeed simply mask the flag. Either is fine by me, what I think is most important for now is that the initramfs generation becomes more deterministic and reproducible.

Reply:
>It's parsing modinfo.

Oh thanks, I completely missed that.

And:
```
$ modinfo nvidia.ko  | grep gsp
firmware:       nvidia/550.78/gsp_tu10x.bin
firmware:       nvidia/550.78/gsp_ga10x.bin
```
So sounds we're ok there, thanks for having looked into it.

Reply:
Not that masking anyway hurts, there's really little advantages to including it even if you want to use fbdev=1, early boot messages can still be displayed by simpledrm,simplefb,efifb, and even nouveau (thing with fbdev=1 is that it can unload other DRM_* like simpledrm and nouveau and doesn't need to be loaded first -- that's why nvidia implemented it in the first place, taking over).

Reply:
Does make me wonder if modules should have a opt-in instead (MODULES_INITRAMFS=1 or so). When not set it'd create the disable files unconditionally (same as USE being disabled), and when enabled it'd add the IUSE.

I think of things like say.. xpadneo, who would need gamepad drivers on a initramfs? And an extra USE that does nothing useful.

Reply:
Not having the IUSE on packages where it's not relevant would make globally setting USE=initramfs bit more sane too.

Reply:
For a small story, had [someone request](https://bugs.gentoo.org/932781) that we add USE=initramfs on nvidia-drivers because it's supposedly the only way it worked for them.

Had me confused until they shown me the error they were getting, turns out dracut didn't include `/etc/modprobe.d/nvidia.conf` which limit access to the nvidia cards to the video group (to reduce security risks). And it allowed things to work for them despite not being in video group.

So, I'm glad adding the USE remained an optional thing :eyes:. Not sure if I'd have realized before exposing users.

Reply:
Hmm, would we run into the same issue with other GPU drivers, or is it just Nvidia?

Reply:
Well, anything that uses modprobe.d files for options could be an issue if it's not included by default (not that I tested this, just assuming given what happened). Not that many modules need these.

Suppose such package could just install a dracut config file of their own to handle this though, but it's easy to overlook if support is added on a whim.